src/imp_pl.rs

```diff
@@ -74,32 +74,76 @@ impl<T> OnceCell<T> {
     /// the contents are acquired by (synchronized to) this thread.
     pub(crate) unsafe fn get_unchecked(&self) -> &T {
         debug_assert!(self.is_initialized());
-        let slot: &Option<T> = &*self.value.get();
-        match slot {
-            Some(value) => value,
-            // This unsafe does improve performance, see `examples/bench`.
-            None => {
-                debug_assert!(false);
-                unreachable_unchecked()
-            }
-        }
+        let slot: &MaybeUninit<T> = &*self.value.get();
+        &*slot.as_ptr()
     }
 
     /// Gets the mutable reference to the underlying value.
     /// Returns `None` if the cell is empty.
     pub(crate) fn get_mut(&mut self) -> Option<&mut T> {
-        // Safe b/c we have a unique access.
-        unsafe { &mut *self.value.get() }.as_mut()
+        if self.is_initialized() {
+            // Safe b/c we have a unique access and value is initialized.
+            unsafe {
+                let slot: &mut MaybeUninit<T> = &mut *self.value.get();
+                Some(&mut *slot.as_mut_ptr())
+            }
+        } else {
+            None
+        }
     }
 
     /// Consumes this `OnceCell`, returning the wrapped value.
     /// Returns `None` if the cell was empty.
-    #[inline]
-    pub(crate) fn into_inner(self) -> Option<T> {
+    pub(crate) fn into_inner(mut self) -> Option<T> {
         // Because `into_inner` takes `self` by value, the compiler statically
         // verifies that it is not currently borrowed.
         // So, it is safe to move out `Option<T>`.
-        self.value.into_inner()
+        //
+        // Safe b/c marking this `OnceCell` as uninitialized below.
+        let value = unsafe { self.take_inner() };
+        if value.is_some() {
+            // `Relaxed` is OK here, because `self` is taking by value, so no
+            // concurrent operations are possible.
+            self.is_initialized.store(false, Ordering::Relaxed);
```

Comment:
Rather than defusing the drop by pretending that the cell is uninitialized, I would prefer to do something like this:

```rust
pub(crate) fn into_inner(mut self) -> Option<T> {
    if !self.is_initialized() { return None; }
    let slot: &MaybeUninit<T> = &*self.value.get();
    let value: T = unsafe { ptr::read(slot.as_ptr()) };
    mem::forget(self);
}

impl<T> Drop for OnceCell<T> {
    fn drop(&mut self) {
        if !self.is_initialized() { return; }
        let slot: &mut MaybeUninit<T> = &mut *self.value.get();
        unsafe { mem::drop_in_place(slot.as_mut_ptr()) };
    }
}
```

It might also make sense to add a couple of private helpers to stop repeating the slot thing everyhere

```rust
impl OnceCell<T> {
    fn as_ptr(&self) -> *const T;
    fn as_ptr_mut(&mut self) -> *mut T;
}
```